### ISSUE TITLE
TV Modifier works only when TV is assigned to a Template

### ISSUE THREAD

**Issue by pmfx** [modxcms/evolution#1218]
In Evo 1.2.1 I have TV called "test_tv" assigned to a Template.
In Template I have this Modifier:

`[*test_tv:ne:then=`test_tv is not empty`:else=`test_tv is empty`*]`

It works fine at this point.

Now, when I uncheck "test_tv" from Template, it doesn't work. 
Value "test_tv is empty" should be returned but there is nothing, like this conditional don't exist.
It worked fine in previous releases (not sure about 1.2).

**Comment by Deesen** [comment]
I guess this is intended and no bug, if it worked before it did not really made sense IMO. I mean why do you want to unlink a TV & template while keeping it available in template? Why not use checkbox "Lock Template Variable for editing" instead?

To achieve what you asked for, Evo would have to query, process, cache all unlinked TVs which will take unnecessary server-resources depending on the amount of old TV-data that may have accumulated over time.

**Comment by pmfx** [comment]
Hey Deesen.
Quick example of use case:

Let's say I have chunk with html header (head,meta,title).
This chunk has modifier:

`[*custom_meta_title:ne:then=`[*custom_meta_title*]`:else=`[*pagetitle*]`*]`

But I don't want some templates to give user this option, yet html header chunk is global, therefore it wont work on templates without this TV.

>To achieve what you asked for, Evo would have to query, process, cache all unlinked TVs

Why MODX would need to do that. It should be done on Modifier level which should check if TV from conditional is assigned to a Template and if not treat is as empty value TV .

**Comment by Deesen** [comment]
In that case I always used a cached IF- or custom-snippet. But now I understand your issue.

**Comment by pmfx** [comment]
You know, the thing is, it worked for a long long time and suddenly it does not. This is unexpected behavior and require site admins to act fast after upgrade cause it may brake site. So it kinda feels like a bug :(

**Comment by Deesen** [comment]
Means with activated phx-plugin it worked, but with internal filters it does not?

**Comment by pmfx** [comment]
It dont work with phx and it dont work with built-in filters.

**Comment by Deesen** [comment]
Just for testing, you can comment out this line, maybe it helps?

https://github.com/modxcms/evolution/blob/develop/manager/includes/document.parser.class.inc.php#L961

**Comment by Deesen** [comment]
With this fix it works for me using PHX and filters, both output correct values.

**Comment by pmfx** [comment]
Your fix works for me too :)

**Comment by Deesen** [comment]
@yama Sorry to bother you but what do you think? Is it a fix or does it lead to new problems?

**Comment by yama** [comment]
@Deesen 
"Variable does not exist" and "Value is empty" is different. In PHP, it is "!Isset" and "empty". I think that it depends on the composition of contents.
```
if(!isset($ph[$key]) && $modifiers) $ph[$key] = '';
```
Is it better to add such an expression? Because it does not have to cause problems with modifier.

**Comment by yama** [comment]
However, it may be problematic to treat unassigned variables as being actually present.
How about treating "nonexistent variables" as "empty values" and adding a new config?

**Comment by yama** [comment]
Or, there is an idea to add special modifiers that can treat "nonexistent variables" as "empty values". It may be the simplest.

**Comment by Deesen** [comment]
I am not sure what you mean with special modifiers. Maybe something like 

    if(!isset($ph[$key]) && $modifiers) {
      $empty = $this->getEmptyPhValue($key, $modifiers);    // new function
      if($empty !== NULL) $ph[$key] = $empty;
    }

**Comment by yama** [comment]
```
[*test_tv:isNull=``:ne:then=test_tv is not empty:else=test_tv is empty*]
```
For example, create a modifier "isNull".
```
if(!isset($ph[$key]) && substr($modifiers,0,6)==='isNull' {
  $ph[$key] = '';
}
```
Or, change this, an already existing "isEmpty" modifier. Just like PHP's "empty".

**Comment by yama** [comment]
Another name is "isNotAssignedTV" and so on.

https://github.com/modxcms/evolution/issues/1218#issuecomment-274818782
However, without thinking hard, maybe Deesen's suggestion might be good.

**Comment by Deesen** [comment]
Me I do not exactly understand what cases are involved / what was originally intended by

    if(!isset($ph[$key]) && !$context) continue;

**Comment by nick0** [comment]
Do you think this issue is related to the above? If no apologies for posting here
On a recently 1.2.1 updated site, I have a lost a "more info" box which has a phx structure like so...
```
[*phx:multitvisempty=`links`:then=`
<p>Etc1...</p>
`:else=`
<p>Etc2...</p>
{{multiTV-links}}
`*]
```
It was working fine in 1.0.15 and all previous.
Any ideas?

**Comment by Deesen** [comment]
@nick0 Maybe it works with my updated document.parser.class.inc.php ?

**Comment by nick0** [comment]
> @nick0 Maybe it works with my updated document.parser.class.inc.php ?

Hmmmm. I have added your file (which fixed the wayfinder and phx issue thank you) but that doesn't seem to work for me in this scenario. I am trying a simplified call like so:
```[*phx:multitvisempty=`slides`:then=`
<p>YES...</p>
`:else=`
<p>NO...</p>
`*]```
It always returns yes regardless of if the MultiTV has content (slides in this test).
Any other tricks?

**Comment by nick0** [comment]
Hi @Deesen, I am experiencing a wierd thing with TVs with your updated document.parser.class.inc.php file and I think this is related to this thread...

I have a tv on one site that has the following options:
```
 Logo, Testimonial, Subscribe Form and sticky Back to Top Button=={{sticky-logo-test}}||
Logo, Anchor Menu, Subscribe Form and sticky Back to Top Button=={{static-logo-anchor}}||
Logo, Sub Menu, Subscribe Form and sticky Back to Top Button=={{static-logo-sub}}||
Logo and sticky Anchor Menu with Back to Top Button=={{sticky-logo-anchor}}||
Logo, Subscribe Form and sticky Back to Top Button=={{sticky-logo}}||
none=={{empty}}
```

This is the selected option:
`Logo, Subscribe Form and sticky Back to Top Button=={{sticky-logo}}||`

The sticky-logo chunk has the following contents
```
{{logo-rhs}}
[*sub-form:is=`rhs`:then=`{{Subscribe-Form-R}}`*]
{{sticky-back-to-top}}
```

As you can see there is an extra tv referenced in this tv:
```[*sub-form:is=`rhs`:then=`{{Subscribe-Form-R}}`*]```

This should display:
```
A logo
A eform form
A sticky back to top button
```

But it does not display the eform form for some reason with the updated document.parser.class.inc.php file.

The TV works as expected when I use the normal "document.parser.class.inc.php" file.

Any ideas on this?

Thanks for all your great work.

**Comment by Deesen** [comment]
I have set up a testing case using the demosite and your above details but I could not recreate your issue yet. Of course you must have `[*sub-form*]` assigned to your template and set to "rhs" (so your issue is not really related to this one).

Do you use different type of TVs than me, maybe there is a bug? This way it works for me like expected:

![1218-1](https://cloud.githubusercontent.com/assets/8569221/22412125/402d39de-e6ab-11e6-9f0d-820be8348d11.png)

**Comment by nick0** [comment]
Thanks @Deesen.

> Of course you must have [*sub-form*] assigned to your template and set to "rhs" 
Yes, that is exactly how it is. It was working fine in 1.0.15. Weird that it does not work with your modified document.parser file.

> (so your issue is not really related to this one).
Woops, so sorry to hijack. Happy to create a new issue if you want.

Hmmm, its a mystery. Not sure if it makes any different but my options are radio buttons like below...
![untitled-1](https://cloud.githubusercontent.com/assets/1768308/22501433/a82cf43e-e8b8-11e6-93c1-82a1ce771872.jpg)

If that doesn't change anything then I have no idea what is causing it. Gremlins :)

**Comment by Deesen** [comment]
I changed TVs to radio-buttons and still could not recreate your issue, I get the form displayed as expected. It is really "gremlish"..

**Comment by nick0** [comment]
Ok, thanks for trying @Deesen, I appreciate the you looking into it. It sure is, its a weird one.

**Comment by Deesen** [comment]
I still would like to help but how? Maybe temporary manager-access could bring some light into it if possible.

**Comment by nick0** [comment]
Thank you very much. Ok, It is a client site so what I will do is create a copy of the site and give you full access to that site. Just sending an email through your website. Thanks again

**Comment by nick0** [comment]
Thanks for checking this out @Deesen. That does indeed seem to work now for that site with the original doc parser file and your modified version.

Do the new modifiers work with MultiTV? If not, disabling PHX and enabling the in built filters is not the solution for most of the other sites I work with as they all use MultiTV with some phx in the templates.
 
When disabling PHX and enabling the in built filters on my master / main test site, I get a totally blank page in the front end (not so backwards compatable). Is there documentation anywhere with example usage anywhere regarding the new modifiers? I think using the new modifiers could break a lot of sites without time consuming manual intervention on every phx call.

**Comment by Deesen** [comment]
As I know MultiTV has its own PHX on board as fallback. You need to test what will work or not, I cannot say for sure.

Blank page of course is a no-go, especially without any error-notice on screen or in system log. But I did not work on modifiers, @yama did it. Unfortunately he is very busy for some weeks.

Regarding documentation, I guess this is the best bet for now: http://www.evolution-docs.com/documentation/designing/modifiers/ taken from https://github.com/modxcms/evolution/issues/623

**Comment by nick0** [comment]
Ah, ok thanks @Deesen. I didn't know that about MultiTV. I just did a very quick test in the row template of a MultiTV photo gallery and it seemed to work fine. Not holding much hope for Jako's previous trick with phx:outer then.

Agreed, blank page without any info is a worry. PHx was always good at showing errors :)

Ok thanks for pointing me to the docs page. I think a lot of example usage and parameters are missing for people who are new to the modifiers.

Cheers for your help.

**Comment by MrSwed** [comment]
revert this https://github.com/modxcms/evolution/commit/12c67aad38636705234439fbb3cda7512f4fb475
for fix #1276
